### PR TITLE
Simplify mining stats

### PR DESCRIPTION
**Pull request by quantumagi**
This PR addresses some issues surrounding the mining statistics by creating a more straight-forward and readable implementation thereof.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs [hunk 1]
```diff
@@ -181,6 +181,27 @@ public virtual void InitializeMining()
             startAfter: TimeSpans.TenSeconds);
         }
 
+        private IFederationMember DetermineExpectedMinerForTimestamp(uint headerTime)
+        {
+            // When blocks are missing for given timestamps we have to determine whom was supposed to mine
+            // by looking at the federation make-up at the time and whom mined last.
+
+            ChainedHeader prevBlockMined = this.consensusManager.Tip;
+            while (prevBlockMined.Header.Time >= headerTime)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs:190]: Shouldn't it just be greater than here (as opposed to greater than or equals to)? Cause if the current slot you want to check matches the chain tip time then we still set it back to previous?

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs [hunk 2]
```diff
@@ -220,70 +241,59 @@ private void GatherMiningStatistics()
...
-                    pubKeyRepresentation = "█████";
+                while (currentHeader.Header.Time > currentSlotTime)
+                    currentHeader = currentHeader.Previous;
 
                 // Mined in this slot?
                 if (currentHeader.Header.Time == currentSlotTime)
                 {
+                    pubKey = this.federationHistory.GetFederationMemberForBlock(currentHeader)?.PubKey;
+
+                    if (pubKey == this.federationManager.CurrentFederationKey?.PubKey)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs:260]: Can maybe put this piece of code in a local method and just pass true or false based on `ProducedBlockInLastRound`

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs [hunk 3]
```diff
@@ -181,6 +181,27 @@ public virtual void InitializeMining()
...
+        private IFederationMember DetermineExpectedMinerForTimestamp(uint headerTime)
+        {
+            // When blocks are missing for given timestamps we have to determine whom was supposed to mine
+            // by looking at the federation make-up at the time and whom mined last.
+
+            ChainedHeader prevBlockMined = this.consensusManager.Tip;
+            while (prevBlockMined.Header.Time >= headerTime)
+                prevBlockMined = prevBlockMined.Previous;
+
+            IFederationMember lastMiner = this.federationHistory.GetFederationMemberForBlock(prevBlockMined);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs:193]:
```suggestion
            IFederationMember minerForBlock = this.federationHistory.GetFederationMemberForBlock(prevBlockMined);
```

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs [hunk 4]
```diff
@@ -181,6 +181,27 @@ public virtual void InitializeMining()
...
+            List<IFederationMember> modifiedFederation = this.federationHistory.GetFederationForBlock(prevBlockMined);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs:194]:
```suggestion
            List<IFederationMember> federationAtBlock = this.federationHistory.GetFederationForBlock(prevBlockMined);
```

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs [hunk 5]
```diff
@@ -181,6 +181,27 @@ public virtual void InitializeMining()
...
+
+            int offset = (int)((headerTime - prevBlockMined.Header.Time) / this.network.ConsensusOptions.TargetSpacingSeconds);
+
+            int lastMinerIndex = modifiedFederation.TakeWhile(m => m.PubKey != lastMiner.PubKey).Count();
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs:198]:
```suggestion
            int minerForBlockIndex = modifiedFederation.TakeWhile(m => m.PubKey != lastMiner.PubKey).Count();
```

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs [hunk 6]
```diff
@@ -220,70 +241,59 @@ private void GatherMiningStatistics()
...
 
             // Iterate mining slots.
-            for (int i = 0; i < maxDepth; i++)
+            for (int i = 0; i < maxDepth; i++, currentSlotTime -= this.network.ConsensusOptions.TargetSpacingSeconds)
             {
-                string pubKeyRepresentation = pubKey.ToString().Substring(0, pubKeyTakeCharacters);
-
-                if (pubKey == this.federationManager.CurrentFederationKey?.PubKey)
-                    pubKeyRepresentation = "█████";
+                while (currentHeader.Header.Time > currentSlotTime)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs:252]:
```suggestion
                // Find the chained header for this slot.
                while (currentHeader.Header.Time > currentSlotTime)
```

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs]: Done

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs]: Done

**Pull request merged**